File: L1_JavaToC/Q3matrixArithmetic.c
#include <stdio.h>
#include <stdlib.h>

// Function to read matrix
void read(int **ar, int r, int c) {
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            printf("Enter element at row %d column %d = ", i, j);
            scanf("%d", &ar[i][j]);
        }
    }
}

// Function to display matrix
void display(int **ar, int r, int c) {
    printf("\n");
    for (int i = 0; i < r; i++) {
        for (int j = 0; j < c; j++) {
            printf("%d ", ar[i][j]);
        }
        printf("\n");
    }
}

// Function to add matrices
void add(int **ar1, int **ar2, int r1, int c1, int r2, int c2) {
    if (r1 == r2 && c1 == c2) {
        int **sum = (int **)malloc(r1 * sizeof(int *));
        for (int i = 0; i < r1; i++)
            sum[i] = (int *)malloc(c1 * sizeof(int));

        for (int i = 0; i < r1; i++) {
            for (int j = 0; j < c1; j++) {
                sum[i][j] = ar1[i][j] + ar2[i][j];
            }
        }
        printf("Sum =\n");
        display(sum, r1, c1);

        for (int i = 0; i < r1; i++) free(sum[i]);
        free(sum);
    } else {
        printf("Sum is not possible!\n");
    }
}

// Function to subtract matrices
void sub(int **ar1, int **ar2, int r1, int c1, int r2, int c2) {
    if (r1 == r2 && c1 == c2) {
        int **diff = (int **)malloc(r1 * sizeof(int *));
        for (int i = 0; i < r1; i++)
            diff[i] = (int *)malloc(c1 * sizeof(int));

        for (int i = 0; i < r1; i++) {
            for (int j = 0; j < c1; j++) {
                diff[i][j] = abs(ar1[i][j] - ar2[i][j]);
            }
        }
        printf("Subtraction =\n");
        display(diff, r1, c1);

        for (int i = 0; i < r1; i++) free(diff[i]);
        free(diff);
    } else {
        printf("Subtraction is not possible!\n");
    }
}

// Function to multiply matrices
void multi(int **ar1, int **ar2, int r1, int c1, int r2, int c2) {
    if (c1 == r2) {
        int **mul = (int **)malloc(r1 * sizeof(int *));
        for (int i = 0; i < r1; i++)
            mul[i] = (int *)calloc(c2, sizeof(int));

        for (int i = 0; i < r1; i++) {
            for (int j = 0; j < c2; j++) {
                for (int k = 0; k < c1; k++) {
                    mul[i][j] += ar1[i][k] * ar2[k][j];
                }
            }
        }
        printf("After multiplication =\n");
        display(mul, r1, c2);

        for (int i = 0; i < r1; i++) free(mul[i]);
        free(mul);
    } else {
        printf("Matrix cannot be multiplied!\n");
    }
}

int main() {
    int r1, c1, r2, c2;

    printf("For matrix 1:\n");
    printf("Enter number of rows: ");
    scanf("%d", &r1);
    printf("Enter number of columns: ");
    scanf("%d", &c1);

    int **matrix1 = (int **)malloc(r1 * sizeof(int *));
    for (int i = 0; i < r1; i++)
        matrix1[i] = (int *)malloc(c1 * sizeof(int));

    read(matrix1, r1, c1);

    printf("For matrix 2:\n");
    printf("Enter number of rows: ");
    scanf("%d", &r2);
    printf("Enter number of columns: ");
    scanf("%d", &c2);

    int **matrix2 = (int **)malloc(r2 * sizeof(int *));
    for (int i = 0; i < r2; i++)
        matrix2[i] = (int *)malloc(c2 * sizeof(int));

    read(matrix2, r2, c2);

    printf("Matrix 1 =\n");
    display(matrix1, r1, c1);

    printf("Matrix 2 =\n");
    display(matrix2, r2, c2);

    // Perform operations
    add(matrix1, matrix2, r1, c1, r2, c2);
    sub(matrix1, matrix2, r1, c1, r2, c2);
    multi(matrix1, matrix2, r1, c1, r2, c2);

    // Free memory
    for (int i = 0; i < r1; i++) free(matrix1[i]);
    free(matrix1);
    for (int i = 0; i < r2; i++) free(matrix2[i]);
    free(matrix2);

    return 0;
}

File: L1_JavaToC/Q3matrixArithmetic.java
import java.util.Scanner;

public class Q3matrixArithmetic{
    static Scanner sc=new Scanner(System.in);
    public static void read(int ar[][],int r,int c){
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                System.out.print("Enter element at row "+i+" column "+j+" = ");
                ar[i][j]=sc.nextInt();
                System.out.println();
            }
        }
    }
    public static void display(int ar[][]){
        System.out.println();
        for(int i=0;i<ar.length;i++){
            for(int j=0;j<ar[i].length;j++){
                System.out.print(ar[i][j]+" ");
                
            }
            System.out.println();
        }
    }
    public static void add(int ar1[][],int ar2[][]){
        if(ar1.length == ar2.length){
            if(ar1[0].length==ar2[0].length){
                int sum[][]=new int[ar1.length][ar1[0].length];
                for(int i=0;i<ar1.length;i++){
                    for(int j=0;j<ar1[i].length;j++){
                        sum[i][j]=ar1[i][j]+ar2[i][j];
                    }
                }
                System.out.print("Sum");
                display(sum);
            }else{
                System.out.println("Sum is not possible !");
            }
        }
        else{

            System.out.println("Sum is not possible !");
        }
    }
    public static void sub(int ar1[][],int ar2[][]){
        if(ar1.length == ar2.length){
            if(ar1[0].length==ar2[0].length){
                int sum[][]=new int[ar1.length][ar1[0].length];
                for(int i=0;i<ar1.length;i++){
                    for(int j=0;j<ar1[i].length;j++){
                        sum[i][j]=Math.abs(ar1[i][j]-ar2[i][j]);
                    }
                }
                System.out.print("Subtraction=");
                display(sum);
            }else{
                System.out.println("Subtraction is not possible !");
            }
        }
        else{

            System.out.println("Subtraction is not possible !");
        }
    }
    public static void multi(int ar1[][],int ar2[][]){
        //ar1 m*n and ar2 n*p then only they can be multiply
        //new matrix will be m*p 
        if(ar1[0].length == ar2.length){
            int r=ar1.length;
            int c=ar2[0].length;
            int mul[][]=new int[r][c];
            for(int i=0;i<r;i++){
                for(int j=0;j<c;j++){
                    for(int k=0;k<c;k++){
                        mul[i][j]+=ar1[i][k]*ar2[k][j];
                    }
                }
            }
            System.out.println("After multiplication=");
            display(mul);
        }
        else{
            System.out.println("Matrix can not be multiplied!");
        }
    }
        public static void main(String arg[]){
        System.out.println("For matrix 1!");
        System.out.print("Enter the no of rows=");
        int r1=sc.nextInt();
        System.out.print("Enter no of columns=");
        int c1=sc.nextInt();
        int matrix1[][]=new int[r1][c1];
        read(matrix1, r1, c1);
        System.out.println("For matrix 2!");
        System.out.print("Enter the no of rows=");
        int r2=sc.nextInt();
        System.out.print("Enter no of columns=");
        int c2=sc.nextInt();
        int matrix2[][]=new int[r2][c2];
        read(matrix2, r2, c2);
        System.out.println("MAtrix 1=");
        display(matrix1);
        System.out.println("MAtrix 2=");
        display(matrix2);
        multi(matrix1, matrix2);
    }
}

